### PR TITLE
Eco 639 create order translation support

### PR DESCRIPTION
**Pull request by liorama**

### CODE REVIEW

File: scripts/src/adaptTranslationCsv.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			return dict;
+		}
+		dict[defaultStr] = translation;
+		return dict;
+	}, {} as { [defaultStr: string]: string });
+
+}
+
+async function addTranslationTo(csv: TranslationData, fromDict: { [defaultStr: string]: string }) {
+	return csv.map(([type, key, defaultStr, __, charLimit]) => {
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: you can:
```
return csv.map(([type, key, defaultStr, __, charLimit]) => ({
	Type: type,
	Key: key,
	Default: defaultStr,
	Translation: fromDict[defaultStr] || defaultStr,
	"Character Limit": charLimit,
}));
```

File: scripts/src/adaptTranslationCsv.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+			Key: key,
+			Default: defaultStr,
+			Translation: fromDict[defaultStr] || defaultStr,
+			"Character Limit": charLimit,
+		};
+	});
+}
+
+function writeCsvDataToFile(data: any[], fileName: string) {
+	const options: ExportCsvOptions = {
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: is the type annotation here necessary? 
the compiler should be ok with it even if you don't explicitly tell it that options is of type `ExportCsvOptions`, as it use duck typing it should be fine

File: scripts/src/adaptTranslationCsv.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		quoteStrings: "\"",
+		decimalseparator: ".",
+		showLabels: true,
+		showTitle: false,
+		useBom: true,
+		useKeysAsHeaders: true,
+	};
+
+	const csvExporter = new ExportToCsv(options);
+	writeFile(fileName, csvExporter.generateCsv(data, true), (err: NodeJS.ErrnoException) => {
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: here also i don't understand why the type of the error is important, it can just be `err => { ...`

File: scripts/src/adaptTranslationCsv.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		if (err) {
+			console.error("Error:", err);
+			return;
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: as you don't have much logic after this if, please remove this `return` and use `else` instead

File: scripts/src/adaptTranslationCsv.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		console.log("CSV saved as", fileName);
+	});
+}
+
+export async function processFile(translationFile: string, fileToTranslate: string, saveAs: string | null = null, rowOffSet: number = 1) {
+	if (!translationFile || !fileToTranslate) {
+		console.error("Both input and output file are required");
+		return;
+	}
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: add a line break

File: scripts/src/adaptTranslationCsv.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	const translationCsv = readFileSync(path(translationFile));
+	const outputCsv = readFileSync(path(fileToTranslate));
+	const parsedTranslationCsv = (csvParse as CsvParse)(translationCsv);
+	const parsedCsvToTranslate = (csvParse as CsvParse)(outputCsv);
+	parsedTranslationCsv.splice(0, rowOffSet);
+	parsedCsvToTranslate.splice(0, rowOffSet);
+	const offerToTranslationDict = getOfferTranslation(parsedTranslationCsv);
+	const translatedData = await addTranslationTo(parsedCsvToTranslate, offerToTranslationDict);
+	if (!saveAs) {
```
Comment (nitzantomer) [inline scripts/src/adaptTranslationCsv.ts]: add a line break

File: scripts/src/admin/translations.ts [hunk 7]
```diff
@@ -88,6 +92,7 @@ function handleIterableItem(key: string, item: any, rowConstructor: RowConstruct
 
 function constructRowsFromArray(keyBase: string, arr: any[], rowConstructor: RowConstructor) {
 	if (!arr) {
+		console.warn(`Empty content for KeyBase ${keyBase}`);
 		return [];
 	}
 	let result: CsvRow[] = [];
```
Comment (nitzantomer) [inline scripts/src/admin/translations.ts]: add a line break

File: scripts/src/admin/translations.ts [hunk 8]
```diff
@@ -100,6 +105,7 @@ function constructRowsFromArray(keyBase: string, arr: any[], rowConstructor: Row
 
 function constructRowsFromObj(keyBase: string, obj: { [key: string]: any }, rowConstructor: RowConstructor) {
 	if (!obj) {
+		console.warn(`Empty content for KeyBase ${keyBase}`);
 		return [];
 	}
 	let result: CsvRow[] = [];
```
Comment (nitzantomer) [inline scripts/src/admin/translations.ts]: add a line break

File: scripts/src/admin/translations.ts [hunk 9]
```diff
@@ -100,6 +105,7 @@ function constructRowsFromArray(keyBase: string, arr: any[], rowConstructor: Row
 
 function constructRowsFromObj(keyBase: string, obj: { [key: string]: any }, rowConstructor: RowConstructor) {
 	if (!obj) {
```
Comment (nitzantomer) [inline scripts/src/admin/translations.ts]: how can `!obj` be false? the function definition doesn't show this is a possibility

File: scripts/src/admin/translations.ts [hunk 10]
```diff
@@ -259,6 +267,9 @@ async function processTranslationData(csvDataRows: TranslationData) {
 	const allOfferContents = await OfferContent.find({ select: ["offerId", "content"] } as FindManyOptions<OfferContent>);
 	const allContentTranslations: OffersTranslation = {};
 	csvDataRows.forEach(([__, csvKey, ___, translation]) => {
+		if (!translation) {
+			return;
+		}
```
Comment (nitzantomer) [inline scripts/src/admin/translations.ts]: add a line break

File: scripts/src/models/translations.ts [hunk 11]
```diff
@@ -11,6 +16,30 @@ export class OfferTranslation extends BaseEntity {
...
+		//  todo add cache
+		const languages = criteria.languages;
+		const offerId = criteria.offerId;
+		const paths = criteria.paths;
+		const query = OfferTranslation.createQueryBuilder("translations");
+		if (languages) {
+			query.where("translations.language IN (:languages)", { languages });
+		}
+		if (offerId) {
+			query.andWhere("translations.offer_id = :offerId", { offerId });
```
Comment (nitzantomer) [inline scripts/src/models/translations.ts]: i'm not sure how `andWhere` works, but if the previous predicate isn't true, then the first `where` was not used. same for the next one.

File: scripts/src/models/translations.ts [hunk 12]
```diff
@@ -21,12 +50,12 @@ export class OfferTranslation extends BaseEntity {
...
-	@Column()
+	@PrimaryColumn()
 	public readonly language!: string;
 
 	@Column()
 	public readonly translation!: string;
 
 	@PrimaryColumn({ name: "offer_id" })
-	private readonly offerId?: string;
+	public readonly offerId?: string;
```
Comment (nitzantomer) [inline scripts/src/models/translations.ts]: yuck, why?

File: scripts/src/public/routes/orders.ts [hunk 13]
```diff
@@ -10,17 +10,35 @@ import {
...
  */
 export const createMarketplaceOrder = async function(req: CreateMarketplaceOrderRequest, res: Response) {
-	const order = await createMarketplaceOrderService(req.params.offer_id, req.context.user!, req.logger);
+	const [availableLanguages, availableTranslations] = await OfferTranslation.getSupportedLanguages({
+		languages: req.acceptsLanguages(),
+		offerId: req.params.offer_id,
+		paths: ["orderTitle", "orderDescription"],
+	});
+	const language = req.acceptsLanguages(availableLanguages); // get the most suitable language for the client
+	const orderTranslations: OrderTranslations = availableTranslations.reduce((dict, translation) => {
```
Comment (nitzantomer) [inline scripts/src/public/routes/orders.ts]: the compiler doesn't infer the return type here?

File: scripts/src/public/services/offer_contents.ts [hunk 14]
```diff
@@ -134,7 +136,18 @@ export function sumCorrectQuizAnswers(offerContent: db.OfferContent, form: strin
...
+	if (acceptsLanguagesFunc && acceptsLanguagesFunc().length) {
+		const [supportedLanguages, availableTranslations] = await OfferTranslation.getSupportedLanguages({
+			paths: ["content"],
+			offerId: offerContent.offerId,
+			languages: acceptsLanguagesFunc(),
+		});
+		const language = acceptsLanguagesFunc(supportedLanguages);
+		const translations = availableTranslations.filter(translation => translation.language === language);
+		translatedContent = translations.length ? translations[0].translation : null;
+	}
```
Comment (nitzantomer) [inline scripts/src/public/services/offer_contents.ts]: add a line break

File: scripts/src/public/services/offers.ts [hunk 15]
```diff
@@ -33,56 +35,77 @@ export interface OfferList {
...
 			.map(async offer => {
-				if ((offerCounts.get(offer.id) || 0) >= offer.cap.per_user) {
-					return null;
-				}
-
-				const content = contents.get(offer.id);
-
-				if (!content) {
-					return null;
+					if ((offerCounts.get(offer.id) || 0) >= offer.cap.per_user) {
```
Comment (nitzantomer) [inline scripts/src/public/services/offers.ts]: no, why did you add more indentation?

File: scripts/src/adaptTranslationCsv.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
+import csvParse = require("csv-parse/lib/sync");
+import { readFileSync, writeFile } from "fs";
+import { path } from "./utils";
+import { CsvParse, TranslationData } from "./admin/translations";
+import { ExportToCsv, Options as ExportCsvOptions } from "export-to-csv";
+
+function getOfferTranslation(inputCsv: TranslationData) {
```
Comment (doodyparizada) [inline scripts/src/adaptTranslationCsv.ts]: whats the return type?

File: scripts/src/adaptTranslationCsv.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+		if (!translation || !translation.length) {
+			return dict;
+		}
+		dict[defaultStr] = translation;
+		return dict;
+	}, {} as { [defaultStr: string]: string });
+
+}
+
+async function addTranslationTo(csv: TranslationData, fromDict: { [defaultStr: string]: string }) {
```
Comment (doodyparizada) [inline scripts/src/adaptTranslationCsv.ts]: We might have a typedef for `[defaultStr: string]: string` StringMap

File: scripts/src/adaptTranslationCsv.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+export async function processFile(translationFile: string, fileToTranslate: string, saveAs: string | null = null, rowOffSet: number = 1) {
+	if (!translationFile || !fileToTranslate) {
+		console.error("Both input and output file are required");
+		return;
+	}
+	const translationCsv = readFileSync(path(translationFile));
+	const outputCsv = readFileSync(path(fileToTranslate));
+	const parsedTranslationCsv = (csvParse as CsvParse)(translationCsv);
+	const parsedCsvToTranslate = (csvParse as CsvParse)(outputCsv);
```
Comment (doodyparizada) [inline scripts/src/adaptTranslationCsv.ts]: Why is `csvParse as CsvParse` needed?

File: scripts/src/adaptTranslationCsv.ts (hunk 4 above)
Comment (liorama) [reply scripts/src/adaptTranslationCsv.ts]: I don't agree, I think this is cleaner.

File: scripts/src/models/translations.ts (hunk 11 above)
Comment (liorama) [reply scripts/src/models/translations.ts]: I looked at the source code. It's better (safer) to have a `where` clause first but the class initializes it anyway so it will work.

File: scripts/src/models/translations.ts (hunk 12 above)
Comment (liorama) [reply scripts/src/models/translations.ts]: `translation.Offer !== Offer' but `translation.OfferId === Offer.id`

File: scripts/src/public/routes/orders.ts [hunk 19]
```diff
@@ -10,18 +10,35 @@ import {
...
 
 export type CreateMarketplaceOrderRequest = Request & {
 	params: {
 		offer_id: string;
 	}
 };
+export type OrderTranslations = {
+	orderTitle: string;
+	orderDescription: string;
+	[index: string]: string;
```
Comment (doodyparizada) [inline scripts/src/public/routes/orders.ts]: @nitzantomer ?
